File: 序号/39. 组合总和.cpp
class Solution {
vector<int> v;
void dfs(int idx,vector<int>& c,int target,vector<vector<int>> & ans){
  if(idx>=c.size()) return;
  int x;
  for(int i=idx;i<c.size();++i){
    x=c[i];
    if (x>target) break;
    v.emplace_back(x);
    if(x==target) ans.emplace_back(v);
    dfs(i,c,target-x,ans);
    v.pop_back();
  }
}
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        sort(candidates.begin(),candidates.end());
        dfs(0,candidates,target,ans);
        return ans;
    }
};

File: 序号/39. 组合总和.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        # 排序后，通过回溯往里塞，超过target就跳过
        candidates.sort()
        n=len(candidates)
        ans=[]
        tmp = []
        def dfs(i,remain):
            if i==n:
                return
            if remain==0:
                ans.append(tmp[::])
                return
            for j in range(i,n):
                if remain<candidates[j]:break
                tmp.append(candidates[j])
                dfs(j,remain-candidates[j])
                tmp.pop()
        dfs(0,target)
        return ans